compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt

```diff
@@ -0,0 +1,200 @@
+                    val height = placeable?.height ?: constraints.minHeight
+
+                    layout(width, height) {
+                        placeable?.placeRelative(0, 0)
+                    }
+                }
+            )
+        } as ComposeContainer
+        window.rootViewController = composeContainer
+        window.makeKeyAndVisible()
+        waitForExpectation { onDraw }
+    }
+
+    private fun keyboardUserInfo(height: Dp, animated: Boolean): Map<Any?, *> {
+        val keyboardHeight = screenSize.height - height
+        val keyboardDuration = if (animated) 0.25 else 0.0
+        return mapOf(
+            UIKeyboardAnimationDurationUserInfoKey to NSNumber(double = keyboardDuration),
+            UIKeyboardFrameEndUserInfoKey to NSValue.valueWithCGRect(
+                CGRectMake(
+                    x = 0.0,
+                    y = keyboardHeight.value.toDouble(),
+                    width = screenSize.width.value.toDouble(),
+                    height = keyboardHeight.value.toDouble()
+                )
+            ),
+            UIKeyboardAnimationCurveUserInfoKey to NSNumber(unsignedInteger = 0uL)
+        )
+    }
+
+    fun showKeyboard(height: Dp = 216.dp, animated: Boolean = true) {
+        NSNotificationCenter.defaultCenter().postNotification(
+            NSNotification(
+                name = UIKeyboardWillShowNotification,
+                `object` = null,
+                userInfo = keyboardUserInfo(height, animated)
+            )
+        )
+        keyboardHeight = height
+        resizeKeyboard(height, animated)
+    }
+
+    fun resizeKeyboard(newHeight: Dp, animated: Boolean = true) {
+        NSNotificationCenter.defaultCenter().postNotification(
+            NSNotification(
+                name = UIKeyboardWillChangeFrameNotification,
+                `object` = null,
+                userInfo = keyboardUserInfo(newHeight, animated)
+            )
+        )
+        keyboardHeight = newHeight
+    }
+
+    fun hideKeyboard(animated: Boolean = true) {
+        NSNotificationCenter.defaultCenter().postNotification(
+            NSNotification(
+                name = UIKeyboardWillChangeFrameNotification,
+                `object` = null,
+                userInfo = keyboardUserInfo(0.dp, animated)
+            )
+        )
+        keyboardHeight = 0.dp
+    }
+
+    private val isIdle: Boolean get() {
+        val hadSnapshotChanges = Snapshot.current.hasPendingChanges()
+        val hasPendingRecomposer = Recomposer.runningRecomposers.value.any {
+            it.hasPendingWork
+        }
+        val isApplyObserverNotificationPending = Snapshot.isApplyObserverNotificationPending
+
+        return !hadSnapshotChanges && !hasPendingRecomposer && !isApplyObserverNotificationPending
+    }
+
+    @OptIn(ExperimentalNativeApi::class)
+    fun waitForIdle(deadline: Duration = 5.seconds): Boolean {
+        val defaultIdleScore = 10
+        var idleScore = defaultIdleScore
```

Comment:
Could you explain what this is for?

Reply:
Sure, we're waiting for some number of cycles without update events to treat application state as idle. If event occurred within this period, we start counter of cycles again.

Reply:
What could cause the idle state to change from idle to non-idle?

Reply:
`isIdle` check a bit flucky, this case is being handled on all other platforms as well, but with slightly different approach. For example freezes may cause it to be true under some circumstances.
Also, some things like postponed animation, or similar stuff may cause this move.

Reply:
Maybe I'm missing something, but can you give an example? All the conditions being checked in `isIdle` refer to internal Compose state. If it's implemented correctly, what could be the cause of flakiness?

Reply:
I was talking about this approach, where it uses one more cicle to ensure system is in idle.
https://github.com/JetBrains/compose-multiplatform-core/blob/77a3ddce223a97268f53b35d2dcb99836bb95b87/compose/ui/ui-test/src/androidMain/kotlin/androidx/compose/ui/test/ComposeIdlingResource.android.kt#L76

It looks like the issue presents only in our environment, where we don't have proper control over compose timer.